### DISCUSSION TITLE
How to Redirect in getStaticProps

### DISCUSSION THREAD

**Discussion by Kiarash-Z** [vercel/next.js#11346] [Help] [answered]
Hi, I need to check the existence of a JWT token to navigate the user to the correct route. I put my token validation stuff in `getStaticProps` function. How can I redirect the user to another route in `getStaticProps` function? I have tried this, but it's not working:

```ts
async (context: NextPageContext): Promise<{ props: { token: string } }> => {
  const { token } = nextCookies(context);
  console.log(context) // {}
  if (!token) {
    if (typeof window === 'undefined') {
      // ❌ I get no 'res' key on context for redirection so this throws an error
      context.res.writeHead(302, { Location: '/phone-get' });
      context.res.end();
    } else {
      /* ❌
        this wouldn't work as well. Because the router is not available here.
        The next router only works on the client side
      */
      Router.replace('/phone-get');
    }
  }

  return {
    props: { token },
  };
};
```

**Comment by aralroca** [comment, marked as answer]
getStaticProps is not a middleware before rendering the page. Is executed on the build step. So I think that is not possible.

**Reply by Kiarash-Z** [reply]
What is the recommended way to do this redirection? Should I do it on the client-side or just replace this function with `getServerSideProps`? 🤔

**Reply by aralroca** [reply]
Yes. If you replace your function to `getServerSideProps`, you can use directly the `res.redirect(301, url)`. 

The `req` and `res` are always existing inside the `getServerSideProps` so you don't need a check like `typeof window === 'undefined'`, because this code is never executed on the client.


The `getStaticProps` neither is executed on the client. It is similar to `getServerSideProps` but is executed in the build step, so this is why you don't have the `req` and `res` of each request. Because is only executed once. It's useful to fetch static things.

**Reply by timneutkens** [reply]
Highly recommend not using getServerSideProps when you can use getStaticProps.

You can do a client-side redirect, as expected on a static website.

Eg:

```jsx
import {useEffect} from 'react'
import {useRouter} from 'next/router'
export async function getStaticProps() {
  return {
    props: {
      hello: 'Hello World'
    }
  }
}

export default (props) => {
  const router = useRouter()
  useEffect(() => {
    if(notloggedincondition) {
      router.push('/login')
    }
  }, [])

  return <h1>Rest of the page</h1>
}
```

**Reply by patricebeutler** [reply]
Is using ```res.redirect(301, url)``` inside of ```getServerSideProps``` working as expected?

While it seems that it works correctly when initially loading the page, it seems to fail on client-side-navigation. On client-side-navigation, the behavior I observe is the following:

1. I navigate (on the client-side) to a page with a redirect inside ```getServerSideProps```
2. ```getServerSideProps``` gets executed and returns status 301
3. The url of the 301 redirect is loaded (but not processed in the browser)
4. Next does a hard page-refresh to the current page (not the redirect-target). I think this is maybe because the target of the redirect wasn't a valid json, which would be expected from```getServerSideProps```.
5. Since we have a hard page refresh now, the server does the redirect, which then works (because it's the same as an initial load).

There were a lot of requests for a simple redirect at this point:

1. /only-authorized.json (redirects to /redirect-target)
2. /login (isn't processed by next)
3. /only-authorized (hard reload)
4. /login (redirect-url set in step 3.)

Is there a way around this issue?

Using ```router.push('/login')``` would be an option. But this results in a flash, since the only-authorized page would render before the redirect. Am I correct?

**Reply by Kiarash-Z** [reply]
@timneutkens is using `getStaticProps` for authentications and validations recommended? because it only gets executed on **build time** and not on every user request to the page. So when a user logs in, it wouldn't update the page. Am I right? redirecting on the client-side using `next/router` would result in a flash(if redirecting from page A to page B, page A will be visibile for the user for a second and then redirection happens to page B), but when using `getServerSideProps` this flash would not happen. Therefore, I think using `getServerSideProps` here is the best solution 🤔

**Reply by aralroca** [reply]
@timneutkens I agree with you, doing the redirect on the client is better for automatic page optimization. But how does it affect for bots as Google bot? Because is a 200 status instead of 301...

**Reply by timneutkens** [reply]
> @timneutkens is using `getStaticProps` for authentications and validations recommended? because it only gets executed on **build time** and not on every user request to the page. So when a user logs in, it wouldn't update the page. Am I right? redirecting on the client-side using `next/router` would result in a flash(if redirecting from page A to page B, page A will be visibile for the user for a second and then redirection happens to page B), but when using `getServerSideProps` this flash would not happen. Therefore, I think using `getServerSideProps` here is the best solution 🤔

User state is generally best loaded client-side using skeletons for example (or a separate route). Eg on zeit.co/dashboard we use skeletons while data is being loaded client-side.

> @timneutkens I agree with you, doing the redirect on the client is better for automatic page optimization. But how does it affect for bots as Google bot? Because is a 200 status instead of 301...

Googlebot does 2 passes, the second pass is using a real browser instance and follows through links.

Overall Google will never be logged into your website though.

**Reply by benvp** [reply]
@timneutkens I'm having a bit trouble on getting things right. I like your suggestion on using SSG and handle the authentication client-side and show skeletons for user specific data. Anyway I have an issue here with a mismatch from the server data and the client rendered data. 

I'm using a `AuthContext` which basically stores an `accessToken` and stores it in a cookie. When the page is refreshed the context restores it from the cookie. Obviously, on the server this cookie is not available, therefore in the first render there would be a mismatch (`accessToken` is undefined on server and defined on the first render client side). This results in HTML error messages.

Do you or anybody else have any suggestions on how to achieve this? Until now I'm still stuck with `getInitialProps` to check on the server if the user is available. Anyway, that way I can't use SSG and have to use SSR für these pages.

```ts
export const AuthProvider: React.FC = ({ children }) => {
  const [authInfo, setAuthInfo] = useState<Partial<AuthInfo> | undefined>(restoreAuthInfo);

  useEffect(() => {
    if (authInfo?.accessToken) {
      login(authInfo.accessToken);
    }
  }, []);

  const login = useCallback((accessToken: string) => {
    cookie.set(ACCESS_TOKEN_COOKIE_NAME, accessToken, { expires: addHours(Date.now(), 10) });
    setAuthInfo({ accessToken });
  }, []);

  const logout = useCallback(() => {
    cookie.remove(ACCESS_TOKEN_COOKIE_NAME);
    cookie.remove(USER_COOKIE_NAME);

    setAuthInfo(undefined);
  }, []);

  const value: AuthContextType = {
    login,
    logout,
    accessToken: authInfo?.accessToken,
    user: authInfo?.user,
  };

  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};

// later I use this HOC to redirect the user if he is not authenticated
export const authenticated = <P extends object = {}, IP = P>(WrappedComponent: NextPage<P, IP>) => {
  const Wrapper: NextPage<P, IP> = props => {
    const router = useRouter();

    // this is undefined on the server (of course)
    // but available directly on the first render
    // because it is fetched from the cookie
    const { accessToken } = useAuth();

    useEffect(() => {
      if (!accessToken) {
        router.replace(ROUTES.LOGIN);
      }
    }, [accessToken]);

    return accessToken ? <WrappedComponent {...props} /> : null;
  };

  return Wrapper;
};

// use the HOC
const SomePage = () => <div>Protected content</div>

export default authenticated(SomePage)

```

**Reply by benvp** [reply]
Nevermind, I figured it out by basically following the example here: https://github.com/zeit/next.js/pull/10973 and https://github.com/zeit/next.js/discussions/10724#discussioncomment-726

**Reply by redreceipt** [reply]
For Googlers, `getStaticProps` supports redirects now

https://nextjs.org/docs/basic-features/data-fetching#getstaticprops-static-generation

<img width="635" alt="Screen Shot 2021-05-27 at 10 40 56 AM" src="https://user-images.githubusercontent.com/2659478/119846490-2e10d300-bed8-11eb-9e43-ab20b4bfc34b.png">

**Reply by goranefbl** [reply]
Does getStaticProps run after build? On every page render, I don't think so.

**Reply by msafi** [reply]
And here's the same code but in text 😆 

```js
async function getStaticProps(context) {
  const res = await fetch('https://...');
  const data = await res.json();

  if (!data) {
    return {
      redirect: {
        destination: '/',
        permanent: false,
      },
    };
  }

  return {
    props: { data },
  };
}
```

**Reply by amritk** [reply]
Wait, if it works now. Why am I seeing

`Error: 'redirect' can not be returned from getStaticProps during prerendering`

**Reply by timfuhrmann** [reply]
> Wait, if it works now. Why am I seeing
> 
> 
> 
> `Error: 'redirect' can not be returned from getStaticProps during prerendering`

Can't have it prerendered - you either have to use fallback `true` or `blocking` and must not return a path behind a redirect from `getStaticPaths`

**Comment by stormfar** [comment]
An extension of this request, is it possible to redirect the user _before_ the page is generated in getStaticProps? I would like interpret the data fetched in getStaticProps and then stop the generation of that page if necessary.

**Reply by aralroca** [reply]
I guess that is not possible because `getStaticProps` runs in "build" time, and earlier is not possible.

**Reply by stormfar** [reply]
But with getStaticPaths fallback, pages are generated on first visit. This is a pity, because even though I can redirect on the client, generated (spam) pages I don't want can negatively impact SEO.

**Reply by MohammadPCh** [reply]
Do you find any solution?

**Reply by AshConnolly** [reply]
Agree with @5tormTrooper.
It would be great to intercept the url before static generation. Without this, pages could be getting duplicated. 

For example if I have a dynamic page that takes the param, say a band name, and statically generates using a query parameter to fetch band information, I could end up generating two pages with the same content.
If a user visits:
`/music/Daft Punk` (becoming `/music/Daft%20Punk`) and `/music/daft-punk` the param is the same, but the page is generated twice. 

Additionally is there any automatic url prettifying available in Next.js?

**Reply by AshConnolly** [reply]
For anyone curious I have this helper function solution to solve the above issue. It redirects to a prettier URL as soon as the page renders:
```js
const redirectIfSpacesInUrl = (path, router) => {
  if (typeof window == 'undefined') return // to ensure it does not break SSR
  if (path.includes('%20')) {
    var split = path.split('/')
    var hyphenatedName = split[split.length - 1].replace(/%20/g, '-').toLowerCase()
    router.replace('/path-name/[slug]', `/path-name/${hyphenatedName}`)
  }
}
```
And used like so:
```js
const Page = props => {
  const router = useRouter() 
  redirectIfSpacesInUrl(router.asPath, router)
  ...
```

**Importantly** as far as I can tell using this method does not render the page twice, nor download the assets twice.

**Comment by MohammadPCh** [comment]


**Reply by brianlovin** [reply]
@MohammadPCh since this was also asked in #12613, it might be good to open your question as a separate issue. It does seem useful to have a single place to discuss abuse prevention or potential rate limiting strategies.

**Comment by IRediTOTO** [comment]
You should redirect in useEffect
You can't do that in getStaticProps (it only run when you type `npm run build`), you can redirect in getServerSideProps, I tried with getServerSideProps but it will make your site load slow :\ so I must use useEffect now.
I am finding a solution better than useeffect too. The protected page still render too fast before redirect.

**Comment by tomaszferens** [comment]
I would also like to have some way to do the redirect on the server when accessing a static page with authentication cookie.

I tried to do client-side redirect in `useEffect`, but that resulted in a flash, where the static page was being displayed before the redirect (and authentication request) happened.

I also tried to do the redirect in a separate JS script tag in HTML head that would run *before* the page could be painted, like this:

```
<html>
    <head>
        <script>
            if (document.cookie.includes('authentication')) {
                window.location.href = '/account';
            }
       </script>
    </head>
    <body>
        <div id="__next"></div>
    </body
</html>
```

unfortunately still, no luck, there was a flash between static page and authenticated page.


I ended up rendering the whole page in a div with `display: none;` style, that waits for authentication to happen. Hope that doesn't affect SEO much.

**Comment by timneutkens** [comment]
Opened a RFC here: https://github.com/vercel/next.js/discussions/14890

**Comment by noghartt** [comment]
Using `next/export`, is there any way to redirect from a route to another?

What I did is just something like it:

```tsx
import type { GetStaticProps } from 'next';

const Join = () => <div></div>;

export const getStaticProps: GetStaticProps = async () => ({
  redirect: {
    destination: 'https://app.woovi.com/join',
    permanent: true,
  },
});

export default Join;
```

**Reply by timfuhrmann** [reply]
No a redirect won't work for a static export since both fallback "false" and "blocking" are not supported, which means "getStaticProps" can only be executed on build and redirects can't be pre-rendered.

**Reply by noghartt** [reply]
> No a redirect won't work for a static export since both fallback "false" and "blocking" are not supported, which means "getStaticProps" can only be executed on build and redirects can't be pre-rendered.

Oh, I get it. I have another way to do it using `useEffect`. Thanks!

Do you know any other way to do it?

**Reply by timfuhrmann** [reply]
Before middleware there was a practice of adding a script to the head to execute the redirect logic even before react js - something like this: 

```html
<Head>
    <script id="auth-redirect">
        {`
            if (!document.cookie || document.cookie.indexOf("authenticated=") === -1) {
                location.replace("yourRedirectUrl")
            }
        `}
    </script>
</Head>
```

I don't know whether this solution suits your problem though.
